File: bun-budget-api/src/middlewares/authenticate.ts
import { Context, error } from "elysia";

import { getCredentialId } from "../services/credential-service";
const authenticationError = (set: Context["set"], message: string) => {
  set.headers["www-authenticate"] = "Basic";
  error(401, { error: `Authentication error: ${message}` });
};

const basicAuthHeaderRegex = /^basic (?<authString>.+)/i;
const basicAuthDecryptedFormatRegex = /(?<username>.+):(?<password>.+)/;

const decryptBase64 = (base64: string) => {
  try {
    const decryptedString = Buffer.from(base64, "base64").toString("utf-8");
    return decryptedString;
  } catch {
    return null;
  }
};

const authenticate = async ({ headers, set }: Context) => {
  const authorizationHeader = headers["Authorization"];

  if (!authorizationHeader) {
    return authenticationError(set, "Authorization header missing");
  }

  const authHeaderRegexResult = basicAuthHeaderRegex.exec(authorizationHeader);

  if (!authHeaderRegexResult) {
    return authenticationError(set, "Invalid authorization header");
  }

  const decryptedAuthorization = decryptBase64(authorizationHeader);

  if (!decryptedAuthorization) {
    return authenticationError(set, "Invalid base64 string");
  }

  const decryptedAuthRegexResult = basicAuthDecryptedFormatRegex.exec(
    decryptedAuthorization,
  );

  if (!decryptedAuthRegexResult) {
    return authenticationError(set, "Invalid credentials format");
  }

  const { username, password } = decryptedAuthRegexResult.groups as {
    username: string;
    password: string;
  };

  const credentialId = await getCredentialId(username, password);

  if (!credentialId) {
    return authenticationError(set, "Invalid credentials");
  }

  return { credentialId };
};

export default authenticate;

File: go-budget-api/internal/middlewares/authenticate.go
package middlewares

import (
	b64 "encoding/base64"
	"fmt"

	"github.com/gin-gonic/gin"
	"github.com/oriser/regroup"
)

var basicAuthRegex = regroup.MustCompile(`^basic (?P<authString>.+)`)

type basicAuthRegexGroups struct {
	AuthString string `regroup:"string"`
}

var basicAuthDecryptedFormatRegex = regroup.MustCompile(`(?<username>.+):(?<password>.+)`)

type basicAuthDecryptedFormatRegexGroups struct {
	Username string `regroup:"string"`
	Password string `regroup:"string"`
}

func Authenticate() gin.HandlerFunc {
	return func(context *gin.Context) {
		authorizationHeader := context.GetHeader("Authorization")

		if authorizationHeader == "" {
			createAuthenticationError(context, "Authorization header missing")
			return
		}

		var basicAuthRegexMatch = &basicAuthRegexGroups{}
		if err := basicAuthRegex.MatchToTarget(authorizationHeader, basicAuthRegexMatch); err != nil {
			createAuthenticationError(context, "Invalid authorization header")
			return
		}

		encryptedAuthString := basicAuthRegexMatch.AuthString

		decodedAuthBytes, err := b64.StdEncoding.DecodeString(encryptedAuthString)
		if err != nil {
			createAuthenticationError(context, "Invalid base64 string")
			return
		}

		decodedAuthString := string(decodedAuthBytes)

		basicAuthDecryptedFormatRegexMatch := &basicAuthDecryptedFormatRegexGroups{}
		if err := basicAuthDecryptedFormatRegex.MatchToTarget(decodedAuthString, basicAuthDecryptedFormatRegexMatch); err != nil {
			createAuthenticationError(context, "Invalid credentials format")
			return
		}

		username := basicAuthDecryptedFormatRegexMatch.Username
		password := basicAuthDecryptedFormatRegexMatch.Password

		credentialId := fmt.Sprintf("%s:%s", username, password)

		context.Set("credentialId", credentialId)

		context.Next()
	}
}

func createAuthenticationError(context *gin.Context, message string) {
	fullMessage := fmt.Sprintf("Authentication error: %s", message)

	context.Header("WWW-Authenticate", "Basic")
	context.JSON(401, gin.H{
		"error": fullMessage,
	})
}
